libutils/proc_manager.c

```diff
@@ -0,0 +1,614 @@
+            id, (intmax_t) pid, *lookup_fd);
+
+        free(id);
+        free(cmd);
+        free(description);
+        free(lookup_fd);
+
+        return false;
+    }
+
+    Subprocess *proc = SubprocessNew(id, cmd, description, pid, input, output, lookup_io);
+    MapInsert(manager->procs_by_id, id, proc);
+    MapInsert(manager->procs_by_pid, &(proc->pid), proc);
+
+    if (lookup_fd != NULL)
+    {
+        MapInsert(manager->procs_by_fd, lookup_fd, proc);
+    }
+
+    return true;
+}
+
+
+Subprocess *ProcManagerGetProcessByPID(ProcManager *manager, pid_t pid)
+{
+    assert(manager != NULL);
+    return (Subprocess*) MapGet(manager->procs_by_pid, &pid);
+}
+
+Subprocess *ProcManagerGetProcessById(ProcManager *manager, const char *id)
+{
+    assert(manager != NULL);
+    return (Subprocess*) MapGet(manager->procs_by_id, id);
+}
+
+Subprocess *ProcManagerGetProcessByFD(ProcManager *manager, int fd)
+{
+    assert(manager != NULL);
+    return (Subprocess*) MapGet(manager->procs_by_fd, &fd);
+}
+
+Subprocess *ProcManagerGetProcessByStream(ProcManager *manager, FILE *stream)
+{
+    assert(manager != NULL);
+    int fd = fileno(stream);
+    return (Subprocess*) MapGet(manager->procs_by_fd, &fd);
+}
+
+
+static inline void SoftRemoveProcFromMaps(ProcManager *manager, Subprocess *proc)
+{
+    assert(manager != NULL);
+    assert(proc != NULL);
+
+    MapRemoveSoft(manager->procs_by_id, proc->id);
+    MapRemoveSoft(manager->procs_by_pid, &(proc->pid));
+
+    int lookup_fd = -1;
+    if (proc->lookup_io == 'i')
+    {
+        lookup_fd = fileno(proc->input);
+        assert(lookup_fd != -1);
+    }
+    else if (proc->lookup_io == 'o')
+    {
+        lookup_fd = fileno(proc->output);
+        assert(lookup_fd != -1);
+    }
+
+    if (lookup_fd != -1)
+    {
+        MapRemoveSoft(manager->procs_by_fd, &lookup_fd);
+    }
+}
+
+Subprocess *ProcManagerPopProcessByPID(ProcManager *manager, pid_t pid)
+{
+    assert(manager != NULL);
+
+    Subprocess *proc = MapGet(manager->procs_by_pid, &pid);
+    SoftRemoveProcFromMaps(manager, proc);
+    return proc;
+}
+
+Subprocess *ProcManagerPopProcessById(ProcManager *manager, const char *id)
+{
+    assert(manager != NULL);
+
+    Subprocess *proc = MapGet(manager->procs_by_id, id);
+    SoftRemoveProcFromMaps(manager, proc);
+    return proc;
+}
+
+Subprocess *ProcManagerPopProcessByFD(ProcManager *manager, int fd)
+{
+    assert(manager != NULL);
+
+    Subprocess *proc = MapGet(manager->procs_by_fd, &fd);
+    SoftRemoveProcFromMaps(manager, proc);
+    return proc;
+}
+
+Subprocess *ProcManagerPopProcessByStream(ProcManager *manager, FILE *stream)
+{
+    assert(manager != NULL);
+
+    int fd = fileno(stream);
+    Subprocess *proc = MapGet(manager->procs_by_fd, &fd);
+    SoftRemoveProcFromMaps(manager, proc);
+    return proc;
+}
+
+
+static inline void RemoveProcFromMaps(ProcManager *manager, Subprocess *proc, int lookup_fd)
```

Comment:
It seems a little backwards / error prone to me that `lookup_fd` is a parameter to this function (separate to the subprocess). `SoftRemoveProcFromMaps()` does not have this parameter.

Reply:
Because `SoftRemoveProcFromMaps()` is not used in the place where the `input/output` streams are closed already and thus unusable for `fileno()`. I'll add a comment.